pkg/providers/ovirt/templates/template-finder.go

```diff
@@ -0,0 +1,86 @@
+package templates
+
+import (
+	"fmt"
+	"strings"
+
+	templatev1 "github.com/openshift/api/template/v1"
+	tempclient "github.com/openshift/client-go/template/clientset/versioned/typed/template/v1"
+	ovirtsdk "github.com/ovirt/go-ovirt"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+const (
+	templateNamespace = "openshift"
+	defaultLinux      = "rhel8"
+	defaultWindows    = "windows"
+)
+
+var osInfo = map[string]string{
+	"Red Hat Enterprise Linux Server": "rhel",
+	// TODO add more
+}
+
+// TemplateFinder attempts to find a template based on given parameters
+type TemplateFinder struct {
+	client *tempclient.TemplateV1Client
+}
+
+// NewTemplateFinder creates new TemplateFinder
+func NewTemplateFinder(tempClient *tempclient.TemplateV1Client) *TemplateFinder {
+	return &TemplateFinder{
+		client: tempClient,
+	}
+}
+
+// FindTemplate attempts to find best match for a template based on the source VM
+func (f *TemplateFinder) FindTemplate(vm *ovirtsdk.Vm) (*templatev1.Template, error) {
+	os := findOperatingSystem(vm)
+	workload := getWorkload(vm)
+	return f.getTemplate(os, workload)
+}
+
+func findOperatingSystem(vm *ovirtsdk.Vm) string {
+	if gos, found := vm.GuestOperatingSystem(); found {
+		distribution, _ := gos.Distribution()
+		version, _ := gos.Version()
+		fullVersion, _ := version.FullVersion()
+		return fmt.Sprintf("%s%s", osInfo[distribution], fullVersion)
+	}
+	if os, found := vm.Os(); found {
+		osType, _ := os.Type()
+		// limit number of possibilities
+		osType = strings.ToLower(osType)
+		if strings.Contains(osType, "linux") || strings.Contains(osType, "rhel") {
+			return defaultLinux
+		} else if strings.Contains(osType, "win") {
+			return defaultWindows
+		}
+	}
+	// return empty to fail label selector
+	return ""
```

Comment:
Shouln't we have a defaultLinux as a fallback? As we have to choose one template?

Reply:
We always will have either `guestOS` or `OS`.  It is here to make compiler happy.